brambling/utils/invites.py

```diff
@@ -0,0 +1,510 @@
+        return self._content
+
+    def get_sender_display(self):
+        if self.invite.user:
+            return self.invite.user.get_full_name()
+        else:
+            raise NotImplementedError
+
+    def accept(self):
+        raise NotImplementedError
+
+    def post_accept_url(self):
+        raise NotImplementedError
+
+    def manage_allowed(self):
+        raise NotImplementedError
+
+    def post_manage_url(self):
+        raise NotImplementedError
+
+    def get_mailer(self):
+        return InviteMailer(
+            site=get_current_site(self.request),
+            secure=self.request.is_secure(),
+            invite=self.invite,
+            content=self.get_content(),
+            key="invite_{}".format(self.slug),
+        )
+
+    def send(self):
+        self.get_mailer().send()
+        self.invite.is_sent = True
+        self.invite.save()
+
+    @classmethod
+    def get_invites(cls, content=None):
+        kwargs = {'kind': cls.slug}
+        if content is not None:
+            kwargs['content_id'] = content.pk
+        return Invite.objects.filter(**kwargs)
+
+    @classmethod
+    def get_or_create(cls, request, email, content):
+        instance, created = Invite.objects.get_or_create_invite(
+            email=email,
+            user=request.user,
+            kind=cls.slug,
+            content_id=content.pk,
+        )
+        invite = cls(request, instance, content)
+        return invite, created
+
+    @classmethod
+    def _get_fake_content(cls):
+        raise NotImplementedError
+
+    @classmethod
+    def get_fake_invite(cls, request):
+        """Returns a fake invite (i.e. for testing emails)"""
+        instance = Invite(
+            email='support@dancerfly.com',
+            user=request.user,
+            kind=cls.slug,
+            content_id=0,
+        )
+        invite = cls(request, instance)
+        invite._content = cls._get_fake_content()
+        return invite
+
+
+@register_invite
+class EventInvite(BaseInvite):
+    model = Event
+    slug = 'event'
+    verbose_name = 'Can attend event'
+    accept_template = 'brambling/invites/event.html'
+
+    def accept(self):
+        # Create an order if one doesn't exist already.
+        # For invite-only events, this is the only way
+        # for people to create an order.
+        Order.objects.for_request(
+            event=self.get_content(),
+            request=self.request,
+            create=True
+        )
+
+    def post_accept_url(self):
+        content = self.get_content()
+        return reverse('brambling_event_shop', kwargs={
+            'event_slug': content.slug,
+            'organization_slug': content.organization.slug
+        })
+
+    def manage_allowed(self):
+        # Invites can be created or modified by people who
+        # can edit the event.
+        event = self.get_content()
+        return self.request.user.has_perm('edit', event)
```

Comment:
That's exactly the direction I think we should go! But it would have made this PR even more massive, and it'll be easy to implement things like that later on.